crates/router/src/core/user.rs

```diff
@@ -1407,6 +1407,237 @@ pub async fn get_user_details_in_merchant_account(
     ))
 }
 
+pub async fn list_user_roles_details(
+    state: SessionState,
+    user_from_token: auth::UserFromToken,
+    request: user_api::GetUserRoleDetailsRequest,
+    _req_state: ReqState,
+) -> UserResponse<Vec<user_api::GetUserRoleDetailsResponseV2>> {
+    let required_user = utils::user::get_user_from_db_by_email(&state, request.email.try_into()?)
+        .await
+        .to_not_found_response(UserErrors::InvalidRoleOperation)?;
+
+    let requestor_role_info = roles::RoleInfo::from_role_id(
+        &state,
+        &user_from_token.role_id,
+        &user_from_token.merchant_id,
+        &user_from_token.org_id,
+    )
+    .await
+    .to_not_found_response(UserErrors::InternalServerError)
+    .attach_printable("Failed to fetch role info")?;
+
+    let user_roles_set: HashSet<_> = state
+        .store
+        .list_user_roles_by_user_id(ListUserRolesByUserIdPayload {
+            user_id: required_user.get_user_id(),
+            org_id: Some(&user_from_token.org_id),
+            merchant_id: (requestor_role_info.get_entity_type() <= EntityType::Merchant)
+                .then_some(&user_from_token.merchant_id),
+            profile_id: (requestor_role_info.get_entity_type() <= EntityType::Profile)
+                .then_some(&user_from_token.profile_id)
+                .cloned()
+                .flatten()
+                .as_ref(),
+            entity_id: None,
+            version: None,
+            status: None,
+            limit: None,
+        })
+        .await
+        .change_context(UserErrors::InternalServerError)
+        .attach_printable("Failed to fetch user roles")?
+        .into_iter()
+        .collect();
+
+    let org_name = state
+        .store
+        .find_organization_by_org_id(&user_from_token.org_id)
+        .await
+        .change_context(UserErrors::InternalServerError)
+        .attach_printable("Org id not found")?
+        .get_organization_name();
+
+    let org = NameIdUnit {
+        id: user_from_token.org_id.clone(),
+        name: org_name,
+    };
+
+    let (merchant_ids, merchant_profile_ids) = user_roles_set.iter().try_fold(
+        (Vec::new(), Vec::new()),
```

Comment:
You can probably use something like `Vec::with_capacity(user_roles_set.len())` here, if the size of the `Vec` is known beforehand, and is same as `user_roles_set`.

Reply:
No the size is not fix, we will be only considering those use roles, where the entity type is merchant and profile.
And ignoring the internal and organisation in this use case

Reply:
No the size is not fix, we will be only considering those use roles, where the entity type is merchant and profile.
And ignoring the internal and organisation in this use case